man/SAPHanaSR.7
.\" Version: 0.162.5
.\"
.TH SAPHanaSR 7 "17 Jan 2026" "" "SAPHanaSR"
.\"
.SH NAME
.\"
SAPHanaSR \- Tools for automating SAP HANA system replication in scale-up setups.
.PP
.\"
.SH DESCRIPTION
.\"
\fBOverview\fR
.PP
The SAPHanaSR package provides resource agents (RA) and tools for setting up
and managing automation of SAP HANA system replication (SR) in scale-up setups.
.pp
System replication will help to replicate the database data from one site to
another site in order to compensate for database failures. With this mode of
operation, internal SAP HANA high-availability (HA) mechanisms and the Linux
cluster have to work together.
.PP
The SAPHana RA performs the actual check of the SAP HANA
database instances and is configured as a multi-state resource.
Managing the two SAP HANA instances means that the resource agent controls
the start/stop of the instances. In addition the resource agent is able to
monitor the SAP HANA databases on landscape host configuration level.
For this monitoring the resource agent relies on interfaces provided by SAP.
As long as the HANA landscape status is not "ERROR" the Linux cluster will not
act. The main purpose of the Linux cluster is to handle the takeover to the
other site. Only if the HANA landscape status indicates that HANA can not
recover from the failure and the replication is in sync, then Linux will act.
.PP
An important task of the resource agent is to check the synchronisation status
of the two SAP HANA databases. If the synchronisation is not "SOK", then the
cluster avoids to takeover to the secondary side, if the primary fails. This is
to improve the data consistency.
.PP
Besides the resource agents, the package contains HADR provider hook scripts.
This hook scripts are called by HANA at specific events. One important event is
when system replication gets lost. The related hook script informs the HA
cluster immediately, without waiting for the next RA monitor. This ensures data
consistency even in corner cases. Another use case for hook scripts is allowing
manual takeover if the HANA resource is in maintenance mode, but blocking manual
takeover otherwise. This prevents from dual-primary situations.
.PP
Note: To automate SAP HANA SR in scale-out setups, please use the package
SAPHanaSR-ScaleOut.
.PP
\fBSituations\fR
.PP
.\" TODO
In order to illustrate the meaning of the above overview, some important
situations are described below. This is not a complete description of all
situations.
.PP
1. \fBSite takeover after primary has failed\fR
.br
This is the basic use case for HANA SR automation.
.br
1. site_A is primary, site_B is secondary - they are in sync.
.br
2. site_A crashes.
.br
3. site_B does the takeover and runs now as new primary. The Linux cluster
always makes sure the node at site_B is shut off before any other action is
initiated.
.PP
2. \fBPrevention against dual-primary\fR
.br
A primary absolutely must never be started, if the cluster does not know
anything about the other site.
On initial cluster start, the cluster needs to detect a valid HANA system
replication setup, including system replication status (SOK) and last primary
timestamp (LPT). This is necessary to ensure data integrity.
.PP
The rational behind this is shown in the following scenario:
.br
1. site_A is primary, site_B is secondary - they are in sync.
.br
2. site_A crashes (remember the HANA ist still marked primary).
.br
3. site_B does the takeover and runs now as new primary.
.br
4. DATA GETS CHANGED ON NODE2 BY PRODUCTION
.br
5. The admin also stops the cluster on site_B (we have two HANAs both
   internally marked down and primary now).
.br
6. What, if the admin would now restart the cluster on site_A?
.br
6.1 site_A would take its own CIB after waiting for the initial fencing
    time for site_B.
.br
6.2 It would "see" its own (cold) primary and the fact that there was a
    secondary.
.br
6.3 It would start the HANA from point of time of step 1.->2. (the crash),
    so all data changed in between would be lost.
.br
This is why the Linux cluster needs to enforce a restart inhibit.
.PP
There are two options to get back both, SAP HANA SR and the Linux cluster,
into a fully functional state:
.br
a) the admin starts both nodes again
.br
b) In the situation where the site_B is still down, the admin starts the
   primary on site_A manually.
.br
The Linux cluster will follow this administrative decision. In both cases the
administrator should register and start a secondary as soon as posible. This
avoids a full log partition with consequence of a DATABASE STUCK.
.PP
3. \fBAutomatic registration as secondary after site failure and takeover\fR
.br
The cluster can be configured to register a former primary database
automatically as secondary. If this option is set, the resource agent
will register a former primary database as secondary during cluster/resource
start.
.PP
4. \fBSite takeover not preferred over local re-start\fR
.br
SAPHanaSR-ScaleOut allows to configure, if you prefer to takeover to the
secondary after the primary landscape fails. The alternative is to restart the
primary landscape, if it fails and only to takeover when no local restart is
possible anymore. This can be tuned by SAPHana(7) parameters.
.br
The current implementation only allows to takeover in case the landscape status
reports 1 (ERROR). The cluster will not takeover, when the SAP HANA still
tries to repair a local failure.
.\" see \fBPREFER_SITE_TAKEOVER\fR
.PP
\fBImplementation\fR
.PP
The two HANA database systems (primary and secondary site) are managed by the
same single Linux cluster. The number of nodes in that single Linux cluster
usually is two.
.PP
The HANA consists of two sites with one node each.
.br
Note: A third HANA site can be added, with another system replication (aka
multi-target replication ). This will be tolerated by the cluster. Nevertheless,
this additional pieces are not managed by the cluster. Therefor in multi-target
setups, the AUTOMATED_REGISTER feature needs the register_secondaries_on_takeover
feature of HANA.
.PP
A common STONITH mechanism is set up for all nodes across all the sites.
.PP
Since the IP address of the primary HANA database system is managed by the
cluster, only that single IP address is needed to access any nameserver
candidate.
.PP
\fBBest Practice\fR
.PP
\fB*\fR Use two independent corosync rings, at least one of them on bonded network.
Resulting in at least three physical links. Unicast is preferred.
.PP
\fB*\fR Use Stonith Block Device (SBD), shared LUNs across all nodes on all
sites. Of course, together with hardware watchdog.
.PP
\fB*\fR Align all timeouts in the Linux cluster with the timeouts of the underlying
infrastructure - particuarly network, storage and multipathing.
.PP
\fB*\fR Check the installation of OS and Linux cluster on all nodes before doing
any functional tests.
.PP
\fB*\fR Carefully define, perform, and document tests for all scenarios that should
be covered, as well as all maintenance procedures.
.PP
\fB*\fR Test HANA features without Linux cluster before doing the overall
cluster tests.
.PP
\fB*\fR Test basic Linux cluster features without HANA before doing the overall
cluster tests.
.PP
\fB*\fR Be patient. For detecting the overall HANA status, the Linux cluster needs
a certain amount of time, depending on the HANA and the configured intervals and
timeouts.
.PP
\fB*\fR Before doing anything, always check for the Linux cluster's idle status,
left-over migration constraints, and resource failures as well as the HANA
landscape status, and the HANA SR status.
.PP
\fB*\fR Manually activating a HANA primary creates risk of a dual-primary situation.
The user is responsible for data integrity. See also susTkOver.py(7).
.PP
.\"
.SH REQUIREMENTS
.\"
For the current version of the package SAPHanaSR, the support is limited to the
following scenarios and parameters:
.PP
1. HANA scale-up cluster with system replication. The two HANA database
systems (primary and secondary site) are managed by one Linux cluster.
The number of nodes in that single Linux cluster is two.
Note: A three-node Linux cluster is possible, but not covered by current
best practices.
.PP
2. Technical users and groups such as sidadm are defined locally in the
Linux system. If users are resolved by remote service, local caching is
necessary. Substitute user (su) to sidadm needs to work reliable and without
customized actions or messages. Supported shell is bash.
.PP
3. Strict time synchronization between the cluster nodes, e.g. NTP. All nodes of
the Linux cluster have configured the same timezone.
.PP
4. For scale-up the following SAP HANA SR scenarios are supported with the
SAPHanaSR package:
.br
	4.1 performance-optimized (memory preload on secondary)
.br
	4.2 cost-optimized (e.g. with QA on secondary)
.br
	4.3 multi-tier with HANA 2.0 (replication chaining, third node connected to second one)
.br
	4.4 multi-target with HANA 2.0 (star replication, third node connected to primary one)
.br
	4.5 single-tenant or multi-tenant (MDC) for all of the above
.br
	4.6 multiple independent HANA SR pairs (MCOS) in one cluster
.br
Note: For MCOS, there must be no constraints between HANA SR pairs.
.PP
5. Only one system replication between the two SAP HANA databases in the Linux cluster.
Maximum one system replication to a HANA database outside the Linux cluster.
See also item 12 below.
.PP
6. The replication mode is either sync or syncmem for the controlled replication.
Replication mode async is not supported. The operation modes delta_datashipping,
logreplay and logreplay_readaccess are supported. The operation mode logreplay
is default.
.PP
7. Both SAP HANA database systems have the same SAP Identifier (SID) and
Instance Number (INO).
.PP
8. Besides SAP HANA you need SAP hostagent installed and started on your system.
For SystemV style, the sapinit script needs to be active.
For systemd style, the services saphostagent and SAP${SID}_${INO} can stay enabled. The systemd enabled saphostagent and instance´s sapstartsrv is supported from SAPHanaSR 0.155 onwards. Please refer to the OS documentation for the systemd version. Please refer to SAP documentation for the SAP HANA version.
Combining systemd style hostagent with SystemV style instance is allowed.
However, all nodes in one Linux cluster have to use the same style.
.PP
9. Automated start of SAP HANA database systems during system boot must be switched
off.
.PP
10. The RAs' monitoring operations have to be active.
.PP
11. Using HADR provider hook for srConnectionChanged() by enabling SAPHanaSR.py
is strongly recommended. This might become mandatory in furture versions.
.PP
12. For scale-up, the current resource agent supports SAP HANA in system
replication beginning with HANA version 1.0 SPS 7 patch level 70,
recommended is SPS 11.
With HANA 2.0 SPS04 and later multi-target system replication
is possible as well. Even in HANA multi-target environments, the current
resource agent manages only two sites. Thus only two HANA sites are part
of the Linux cluster. See also item 5 above.
.PP
13. Colocation constraints between the SAPHanaController or SAPHana RA and
other resources are allowed only if they do not affect the RA's scoring.
The location scoring finally depends on system replication status and must not
be over-ruled by additional constraints. Thus it is not allowed to define rules
forcing a SAPHanaController or SAPHana master to follow another resource.
.PP
14. The Linux cluster needs to be up and running to allow HA/DR provider events
being written into CIB attributes. The current HANA SR status might differ
from CIB srHook attribute after cluster maintenance.
.PP
15. Once a HANA system replication site is known to the Linux cluster, that
exact site name has to be used whenever the site is registered manually. At any
time only one site is configured as primary replication source.
.PP
16. Reliable access to the /hana/shared/ filesystem is crucial for HANA and the
Linux cluster.
.PP
17. HANA feature Secondary Time Travel is not supported.
.PP
18. In MDC configurations the HANA database is treated as a single system
including all database containers. Therefor, cluster takeover decisions are
based on the complete status independent of the status of individual containers.
.PP
19. If a third HANA site is connected by system replication, that HANA is not
controlled by another SUSE HA cluster. If that third site should work as part
of a fall-back HA cluster in DR case, that HA cluster needs to be in standby.
.PP
20. RA and srHook runtime almost completely depends on call-outs to controlled
resources, OS and Linux cluster. The infrastructure needs to allow these call-outs
to return in time.
.PP
21. The SAP HANA Fast Restart feature on RAM-tmfps as well as HANA on persistent
memory can be used, as long as they are transparent to SUSE HA.
.PP
22. The SAP HANA hostname or virtual hostname should follow RFC-952.
.PP
23. The SAP HANA site name is from 2 up to 32 characters long. It starts with a
character or number. Subsequent characters may contain dash and underscore.
However, underscore should be avoided.
.PP
24. The SAPHanaController RA, the SUSE HA cluster and several SAP components
need read/write access and sufficient space in the Linux /tmp filesystem.
.PP
25. SAP HANA Native Storage Extension (NSE) is supported.
Important is that this feature does not change the HANA topology or interfaces.
In opposite to Native Storage Extension, the HANA Extension Nodes are changing
the topology and thus currently are not supported. 
Please refer to SAP documentation for details.
.PP
26. The Linux user root´s shell is /bin/bash, or completely compatible.
.PP
27. No manual actions must be performed on the HANA database while it is controlled
by the Linux cluster. All administrative actions need to be aligned with the cluster.
See also SAPHanaSR_maintenance_examples(7).
.PP
28. No firewall rules must block any needed port. For a potential local firewall,
see also firewalld(8).
.PP
29. No SELinux rules must block any needed action. See also selinux(8).
.PP
30. The Linux cluster should not start automatically at reboot on a node that 
recently has been fenced.
.PP
31. Sizing of both HANA sites needs to be done according to SAP rules. The scale-up
scenarios require both sites prepared for running the primary HANA database.
.PP
.\"
.SH BUGS
.\"
In case of any problem, please use your favourite SAP support process to open
a request for the component BC-OP-LNX-SUSE.
.\" TODO Please report any other feedback and suggestions to feedback@suse.com.
.PP
.\"
.SH SEE ALSO
.\"
\fBocf_suse_SAPHanaTopology\fP(7) , \fBocf_suse_SAPHana\fP(7) , \fBocf_heartbeat_IPaddr2\fP(7) ,
\fBSAPHanaSR_basic_cluster\fP(7) ,
\fBSAPHanaSR.py\fP(7) , \fBsusCostOpt.py\fP(7) , \fBsusTkOver.py\fP(7) , \fBsusChkSrv.py\fP (7) ,
\fBSAPHanaSR-monitor\fP(8) , \fBSAPHanaSR-showAttr\fP(8) ,
\fBntp.conf\fP(5) , \fBchrony.conf\fP(5) , \fBsystemctl\fP(1) ,
\fBstonith\fP(8) , \fBsbd\fP(8) , \fBstonith_sbd\fP(7) , \fBstonith_admin\fP(8) ,
\fBcrm\fP(8) , \fBcorosync.conf\fP(5) , \fBcrm_no_quorum_policy\fP(7) ,
\fBsaptune\fP(8) , \fBcs_show_hana_info\fP(8) , \fBsupportconfig\fP(8) ,
.br
https://documentation.suse.com/sbp/sap/ ,
.br
https://documentation.suse.com/sles-sap/ ,
.br
https://www.suse.com/releasenotes/ ,
.br
https://www.susecon.com/doc/2015/sessions/TUT19921.pdf ,
.br
https://www.susecon.com/doc/2016/sessions/TUT90846.pdf ,
.br
https://www.suse.com/media/presentation/TUT90846_towards_zero_downtime%20_how_to_maintain_sap_hana_system_replication_clusters.pdf ,
.br
http://scn.sap.com/community/hana-in-memory/blog/2014/04/04/fail-safe-operation-of-sap-hana-suse-extends-its-high-availability-solution ,
.br
http://scn.sap.com/docs/DOC-60334 ,
.br
http://scn.sap.com/community/hana-in-memory/blog/2015/12/14/sap-hana-sps-11-whats-new-ha-and-dr--by-the-sap-hana-academy ,
.br
https://wiki.scn.sap.com/wiki/display/ATopics/HOW+TO+SET+UP+SAPHanaSR+IN+THE+COST+OPTIMIZED+SAP+HANA+SR+SCENARIO+-+PART+I
.br
https://blogs.sap.com/2020/01/30/sap-hana-and-persistent-memory/
.PP
.SH AUTHORS
.\"
A.Briel, F.Herschel, L.Pinne.
.PP
.\"
.SH COPYRIGHT
.\"
(c) 2015-2018 SUSE Linux GmbH, Germany.
.br
(c) 2019-2026 SUSE LLC
.br
The package SAPHanaSR comes with ABSOLUTELY NO WARRANTY.
.br
For details see the GNU General Public License at
http://www.gnu.org/licenses/gpl.html
.\"


srHook/SAPHanaSR.py
"""
# SAPHana
# Author:       Fabian Herschel, 2015
# License:      GNU General Public License (GPL)
# Copyright:    (c) 2015-2016 SUSE Linux GmbH
# Copyright:    (c) 2017-2026 SUSE LLC
"""
try:
    from hdb_ha_dr.client import HADRBase
except ImportError as e:
    print("Module HADRBase not found - running outside of SAP HANA? - {0}".format(e))
import os

"""
To use this HA/DR hook provide please add the following lines to your global.ini:
    [ha_dr_provider_SAPHanaSR]
    provider = SAPHanaSR
    path = /usr/share/SAPHanaSR
    execution_order = 1

    [trace]
    ha_dr_saphanasr = info
"""
fhSRHookVersion = "0.162.0"


try:
    class SAPHanaSR(HADRBase):

        def __init__(self, *args, **kwargs):
            # delegate construction to base class
            super(SAPHanaSR, self).__init__(*args, **kwargs)
            self.tracer.info("SAPHanaSR init()")

        def about(self):
            return {"provider_company": "SUSE",
                    "provider_name": "SAPHanaSR",  # class name
                    "provider_description": "Inform Cluster about SR state",
                    "provider_version": "1.0"}

        def srConnectionChanged(self, ParamDict, **kwargs):
            """ finally we got the srConnection hook :) """
            method = "srConnectionChanged"
            self.tracer.info("SAPHanaSR {0} {1}.srConnectionChanged method called with Dict={2}".format(fhSRHookVersion, self.__class__.__name__, ParamDict))
            # self.tracer.info("SAPHanaSR (%s) %s.srConnectionChanged method called with Dict=%s" % (fhSRHookVersion, self.__class__.__name__, ParamDict))
            # myHostname = socket.gethostname()
            # myDatebase = ParamDict["database"]
            mySystemStatus = ParamDict["system_status"]
            mySID = os.environ.get('SAPSYSTEMNAME')
            mysid = mySID.lower()
            myInSync = ParamDict["is_in_sync"]
            myReason = ParamDict["reason"]
            mySite = ParamDict["siteName"]
            self.tracer.info("SAPHanaSR {0}.srConnectionChanged mySystemStatus={1} mySID={2} myInSync={3} myReason={4}".format(self.__class__.__name__, mySystemStatus, mySID, myInSync, myReason))
            if mySystemStatus == 15:
                mySRS = "SOK"
            else:
                if myInSync:
                    # ignoring the SFAIL, because we are still in sync
                    self.tracer.info("SAPHanaSR (%s) %s.srConnectionChanged ignoring bad SR status because of is_in_sync=True (reason=%s)" % (fhSRHookVersion, self.__class__.__name__, myReason))
                    mySRS = ""
                else:
                    mySRS = "SFAIL"
            if mySRS == "":
                myMSG = "### Ignoring bad SR status because of is_in_sync=True ###"
                self.tracer.info("{0}.{1}() {2}\n".format(self.__class__.__name__, method, myMSG))
            elif mySite == "":
                myMSG = "### Ignoring bad SR status because of empty site name in call params ###"
                self.tracer.info("{0}.{1}() {2}\n".format(self.__class__.__name__, method, myMSG))
            else:
                myCMD = "sudo /usr/sbin/crm_attribute -n hana_%s_site_srHook_%s -v %s -t crm_config -s SAPHanaSR" % (mysid, mySite, mySRS)
                rc = os.system(myCMD)
                myMSG = "CALLING CRM: <{0}> rc={1}".format(myCMD, rc)
                self.tracer.info("{0}.{1}() {2}\n".format(self.__class__.__name__, method, myMSG))
                fallback_file_name = "../.crm_attribute.{0}".format(mySite)
                fallback_stage_file_name = "../.crm_attribute.stage.{0}".format(mySite)
                if rc == 0:
                    # cluster attribute set was successfull - delete pending fallback file, if existing
                    try:
                        os.remove(fallback_file_name)
                        self.tracer.info("new event - pending fallback file {0} deleted".format(fallback_file_name))
                    except FileNotFoundError:
                        pass
                else:
                    #
                    # FALLBACK
                    # sending attribute to the cluster failed - using fallback method and write status to a file - RA to pick-up the value during next SAPHanaController monitor operation
                    #
                    myMSG = "sending attribute to the cluster failed - using local file as fallback"
                    self.tracer.info("{0}.{1}() {2}\n".format(self.__class__.__name__, method, myMSG))
                    #
                    # cwd of hana is /hana/shared/<SID>/HDB00/<hananode> we use a relative path to cwd this gives us a <sid>adm permitted directory
                    #     however we go one level up (..) to have the file accessible for all SAP HANA swarm nodes
                    #
                    with open(fallback_stage_file_name, "w", encoding='utf-8') as fallbackFileObject:
                        fallbackFileObject.write("hana_{0}_site_srHook_{1} = {2}".format(mysid, mySite, mySRS))
                    #
                    # release the stage file to the original name (move is used to be atomic)
                    #      .crm_attribute.stage.<site> is renamed to .crm_attribute.<site>
                    #
                    os.rename(fallback_stage_file_name, fallback_file_name)
            return 0
except NameError as e:
    print("Could not find base class ({0})".format(e))
